cmd/lotus-shed/indexes.go

```diff
@@ -879,3 +900,339 @@ var backfillTxHashCmd = &cli.Command{
+		ctx := lcli.ReqContext(cctx)
+
+		basePath, startHeight, err := parsePruneArgs(ctx, cctx, api)
+		if err != nil {
+			return xerrors.Errorf("error parsing prune args: %w", err)
+		}
+
+		if err := pruneEventsIndex(ctx, basePath, startHeight); err != nil {
+			return xerrors.Errorf("error pruning events index: %w", err)
+		}
+
+		return nil
+	},
+}
+
+// pruneAllIndexesCmd is a command that prunes all the indexes
+var pruneAllIndexesCmd = &cli.Command{
+	Name:  "prune-all-indexes",
+	Usage: "Prune all the indexes DBs (msgindex, txhash, events) older than a given height",
+	Flags: []cli.Flag{
+		&cli.IntFlag{
+			Name:  "older-than",
+			Value: 0,
+			Usage: "height to start pruning events older than this epoch, 0 means start from head",
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		srv, err := lcli.GetFullNodeServices(cctx)
+		if err != nil {
+			return err
+		}
+		defer srv.Close() //nolint:errcheck
+
+		api := srv.FullNodeAPI()
+		ctx := lcli.ReqContext(cctx)
+
+		basePath, startHeight, err := parsePruneArgs(ctx, cctx, api)
+		if err != nil {
+			return xerrors.Errorf("error parsing prune args: %w", err)
+		}
+
+		var g = new(errgroup.Group)
+
+		g.Go(func() error {
+			if err := pruneMsgIndex(ctx, basePath, startHeight); err != nil {
+				return xerrors.Errorf("error pruning msgindex: %w", err)
+			}
+
+			return nil
+		})
+
+		g.Go(func() error {
+			if err := pruneTxIndex(ctx, api, basePath, startHeight); err != nil {
+				return xerrors.Errorf("error pruning txindex: %w", err)
+			}
+
+			return nil
+		})
+
+		g.Go(func() error {
+			if err := pruneEventsIndex(ctx, basePath, startHeight); err != nil {
+				return xerrors.Errorf("error pruning events index: %w", err)
+			}
+
+			return nil
+		})
+
+		return g.Wait()
+	},
+}
+
+// pruneEventsIndex is a helper function that prunes the events index.db for a number of epochs starting from a specified height
+func pruneEventsIndex(_ context.Context, basePath string, startHeight int64) error {
+	log.Infof("pruning events index")
+
+	dbPath := path.Join(basePath, databaseType, filter.DefaultDbFilename)
+	db, err := sql.Open(databaseDriver, dbPath+"?_txlock=immediate")
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err := db.Close()
+		if err != nil {
+			fmt.Printf("ERROR: closing db: %s", err)
+		}
+	}()
+
+	tx, err := db.Begin()
+	if err != nil {
+		return err
+	}
+
+	rollback := func(tableName string) {
+		if err := tx.Rollback(); err != nil {
+			fmt.Printf("ERROR: rollback %s tx :%s", tableName, err)
+		}
+	}
+
+	res, err := tx.Exec(deleteEventFromStartHeight, startHeight)
+	if err != nil {
+		rollback(eventTable)
+		return xerrors.Errorf("error deleting event: %w", err)
+	}
+
+	var totalRowsAffected int64
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("error getting rows affected: %w", err)
+	}
+
+	totalRowsAffected += rowsAffected
+	log.Infof("pruned %s table, rows affected: %d", eventTable, rowsAffected)
```

Comment:
not critical: I suggest keeping track of these numbers separately and then logging _after_ you've committed the transaction, because until you do, they're not really gone; if you have a rollback then these log messages will be incorrect

Reply:
if you're concerned about progress logging, print a message before each execution: `log.Infof("pruning `%s` table", eventTable)`; then at least the user knows its working if this takes a long time for some reason (it shouldn't unless it's on a really slow disk)

Reply:
Updated.